src/recvsend.rs

```diff
@@ -338,15 +343,19 @@ where
         futures_util::ready!(self.framed.poll_ready_unpin(cx).map_err(Error::Send))
             .map_err(into_io_error)?;
 
-        let mut guard = self.shared.lock().unwrap();
-
-        let response = match guard.message.take() {
-            Some(response) => response,
-            None => {
-                guard.waker = Some(cx.waker().clone());
-                drop(guard);
-
-                return Poll::Pending;
+        let response = {
+            let mut guard = self.shared.lock().unwrap();
+
+            match guard.message.take() {
+                Some(response) => response,
+                None => {
+                    return if guard.waker.replace(cx.waker().clone()).is_none() {
+                        // Responder was dropped before sending any value, nothing else left to do here.
```

Comment:
Should this return an error instead? I'd consider not sending a response a bug in the code that handles the incoming request.

Reply:
This will drop the stream which causes a reset so it will fail for the other party. I think it should fail on our end too!